data/posts/split-route-modules.md

```diff
@@ -0,0 +1,137 @@
+
+To visualize this as a timeline:
+
+```
+Get Route Module:  |--=======|
+Run clientLoader:            |-----|
+Render:                            |-|
+```
+
+The external API call made from the `clientLoader` is delayed waiting for the hypothetical `MassiveComponent` to download. Since we can't render the component without the data from the client loader, the user has to wait for this network waterfall to complete before the page is rendered.
+
+Ideally we’d like to be able to download the `clientLoader` export independently and run it as soon as it’s available:
+
+```
+Get clientLoader:  |--|
+Get Component:     |=======|
+Run clientLoader:     |-----|
+Render:                     |-|
+```
+
+At a framework level, the easiest way for us to solve this would be to force you to author your route in multiple files (`route/clientLoader.ts`, `route/component.tsx`, etc.) — but we really didn't want to give up on the convenience of the Route Module API. The question is, how do we achieve this?
+
+## Splitting the Route Module
+
+What if the React Router Vite plugin could automatically split these route module exports into multiple smaller modules during the production build?
+
+With the `future.unstable_splitRouteModules` flag enabled, this is exactly what happens.
+
+Using our previous example, our singular route module would end up being split into two separate [virtual modules](https://vite.dev/guide/api-plugin#virtual-modules-convention) — one for the client loader and one for the component.
+
+```ts
+// route.tsx?route-chunk=clientLoader
+export async function clientLoader() {
+  return await fetch("https://example.com/api").then((response) =>
+    response.json(),
+  );
+}
+```
+
+```ts
+// route.tsx?route-chunk=main
+import { MassiveComponent } from "~/components";
+
+export default function Component({ loaderData }) {
+  return <MassiveComponent data={loaderData} />;
+}
+```
+
+Since these exports have now been split into separate modules, the React Router Vite plugin can ensure that they can be downloaded independently.
+
+This optimization is even more pronounced when more of the Route Module API is used. For example, when using `clientLoader`, `clientAction` and `HydrateFallback`, the timeline for a single route module during a client-side navigation might look like this:
+
+```
+Get Route Module:     |--~~~++++=======|
+Run clientLoader:                      |-----|
+Render:                                      |-|
+```
+
+This would instead be optimized to the following:
+
+```
+Get clientLoader:     |--|
+Get clientAction:     |~~~|
+Get HydrateFallback:  SKIPPED
+Get Component:        |=======|
+Run clientLoader:        |-----|
+Render:                        |-|
```

Comment:
Is this missing a `Run clientAction` step?

Reply:
No, the `clientAction` download is just preloading in this case.